acvm/src/pwg/mod.rs

```diff
@@ -174,151 +172,122 @@ impl<B: BlackBoxFunctionSolver> ACVM<B> {
-                            _ => unreachable!("Brillig resolution for non brillig opcode"),
-                        };
-                        let hash = canonical_brillig_hash(&brillig);
-                        self.pending_brillig_label_maps.insert(hash, *opcode_label);
-                        self.pending_foreign_calls.push(UnresolvedBrilligCall {
-                            brillig,
-                            foreign_call_wait_info: oracle_wait_info,
-                        })
-                    }
-                    Ok(OpcodeResolution::Stalled(not_solvable)) => {
-                        if opcode_not_solvable.is_none() {
-                            // we keep track of the first unsolvable opcode
-                            opcode_not_solvable = Some(not_solvable);
-                        }
-                        // We push those opcodes not solvable to the back as
-                        // it could be because the opcodes are out of order, i.e. this assignment
-                        // relies on a later opcodes' results
-                        unresolved_opcodes.push((opcode.clone(), *opcode_label));
-                    }
-                    Err(OpcodeResolutionError::OpcodeNotSolvable(_)) => {
-                        unreachable!("ICE - Result should have been converted to GateResolution")
-                    }
-                    Err(error) => return self.fail(error),
+                    res => res,
                 }
             }
-
-            // Before potentially ending execution, we must save the list of opcodes which remain to be solved.
-            std::mem::swap(&mut self.opcodes_and_labels, &mut unresolved_opcodes);
-
-            // We have oracles that must be externally resolved
-            if self.get_pending_foreign_call().is_some() {
-                return self.status(ACVMStatus::RequiresForeignCall);
+            Opcode::Block(_) | Opcode::ROM(_) | Opcode::RAM(_) => {
+                panic!("Block, ROM and RAM opcodes are not supported by stepwise ACVM")
             }
-
-            // We are stalled because of an opcode being bad
-            if stalled && !self.opcodes_and_labels.is_empty() {
-                let error = OpcodeResolutionError::OpcodeNotSolvable(
-                    opcode_not_solvable
-                        .expect("infallible: cannot be stalled and None at the same time"),
-                );
-                return self.fail(error);
+        };
+        match resolution {
+            Ok(OpcodeResolution::Solved) => {
+                self.instruction_pointer += 1;
+                if self.instruction_pointer == self.opcodes.len() {
+                    self.status(ACVMStatus::Solved)
+                } else {
+                    self.status(ACVMStatus::InProgress)
+                }
+            }
+            Ok(OpcodeResolution::InProgress) => {
+                unreachable!("Opcodes_and_labels should be immediately solvable");
```

Comment:
Can we add more context to this message? Its not immediate clear what "Opcodes_and_labels" means here if looking at it from a dependent package like Noir

Reply:
Addressed in https://github.com/noir-lang/acvm/pull/400